File: assignment 5/python/programs/5.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    # Insert a node at the beginning of the doubly linked list
    def insert_at_beginning(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

    # Insert a node at the end of the doubly linked list
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = new_node
            new_node.prev = current

    # Insert a node after a specific element
    def insert_after(self, data, key):
        new_node = Node(data)
        if self.head is None:
            print("List is empty. Cannot perform insertion.")
            return
        current = self.head
        while current:
            if current.data == key:
                new_node.next = current.next
                if current.next is not None:
                    current.next.prev = new_node
                current.next = new_node
                new_node.prev = current
                return
            current = current.next
        print(f"Element {key} not found in the list.")

    # Delete a node by value
    def delete(self, data):
        if self.head is None:
            print("List is empty. Cannot perform deletion.")
            return
        current = self.head

        while current:
            if current.data == data:
                if current.prev is not None:
                    current.prev.next = current.next
                else:
                    self.head = current.next

                if current.next is not None:
                    current.next.prev = current.prev
                return
            current = current.next

        print(f"Element {data} not found in the list.")

    # Display the doubly linked list
    def display(self):
        if self.head is None:
            print("Doubly Linked List is empty.")
            return

        current = self.head
        while current is not None:
            print(current.data, end=" <-> ")
            current = current.next
        print("None")


if __name__ == "__main__":
    doubly_linked_list = DoublyLinkedList()

    # Insert elements into the doubly linked list
    doubly_linked_list.insert_at_end(1)
    doubly_linked_list.insert_at_end(2)
    doubly_linked_list.insert_at_end(3)

    # Display the doubly linked list
    print("Doubly Linked List:")
    doubly_linked_list.display()

    # Insert an element after a specific element
    doubly_linked_list.insert_after(4, 2)

    # Display the doubly linked list after insertion
    print("Doubly Linked List after Insertion:")
    doubly_linked_list.display()

    # Delete an element from the doubly linked list
    doubly_linked_list.delete(2)

    # Display the doubly linked list after deletion
    print("Doubly Linked List after Deletion:")
    doubly_linked_list.display()

File: assignment 5/c/programs/5.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure for a node in the doubly linked list
struct Node
{
    int data;
    struct Node *prev;
    struct Node *next;
};

// Function to create a new node
struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    if (newNode == NULL)
    {
        printf("Memory allocation failed.\n");
        exit(1);
    }
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

// Function to insert a node at the end of the doubly linked list
struct Node *insertAtEnd(struct Node *head, int data)
{
    struct Node *newNode = createNode(data);

    if (head == NULL)
    {
        return newNode;
    }

    struct Node *current = head;
    while (current->next != NULL)
    {
        current = current->next;
    }

    current->next = newNode;
    newNode->prev = current;
    return head;
}

// Function to delete a node at a specific position in the doubly linked list
struct Node *deleteAtPosition(struct Node *head, int position)
{
    if (head == NULL)
    {
        printf("List is empty.\n");
        return NULL;
    }

    if (position <= 0)
    {
        printf("Invalid position.\n");
        return head;
    }

    if (position == 1)
    {
        struct Node *temp = head;
        head = head->next;
        if (head != NULL)
        {
            head->prev = NULL;
        }
        free(temp);
        return head;
    }

    struct Node *current = head;
    int count = 1;

    // Find the node to be deleted
    while (count < position && current != NULL)
    {
        current = current->next;
        count++;
    }

    if (current == NULL)
    {
        printf("Position out of range.\n");
        return head;
    }

    if (current->prev != NULL)
    {
        current->prev->next = current->next;
    }

    if (current->next != NULL)
    {
        current->next->prev = current->prev;
    }

    free(current);
    return head;
}

// Function to display the doubly linked list
void display(struct Node *head)
{
    if (head == NULL)
    {
        printf("List is empty.\n");
        return;
    }

    printf("Doubly Linked List: ");
    while (head != NULL)
    {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

int main()
{
    struct Node *head = NULL;

    // Initialize a predefined doubly linked list
    head = insertAtEnd(head, 10);
    head = insertAtEnd(head, 20);
    head = insertAtEnd(head, 30);

    printf("Initial Doubly Linked List: ");
    display(head);

    int data, position;

    // Insertion
    printf("\nEnter data to insert at the end: ");
    scanf("%d", &data);
    head = insertAtEnd(head, data);

    // Display the list after insertion
    display(head);

    // Deletion
    printf("\nEnter position to delete: ");
    scanf("%d", &position);
    head = deleteAtPosition(head, position);

    // Display the list after deletion
    display(head);

    return 0;
}
